### DISCUSSION TITLE
Previewer is not working

### DISCUSSION THREAD

**Discussion by BlankWhite** [AvaloniaUI/Avalonia#13667] [Q&A] [answered]
![Snipaste_2023-11-20_08-49-24](https://github.com/AvaloniaUI/Avalonia/assets/11631255/b54e5d7c-b96a-45f0-89fc-05e417a693d5)
my code
https://github.com/BlankWhite/SQLConverter

**Comment by timunie** [comment]
@BlankWhite when does it happen? do you click anything on during previewer? 

In general, please fill the issue template if you file a bug. Thank you.

But for me this is more or less a discussion. I'll convert it for now, hope you don't mind. 

-------

Looking roughly into your sample, I see that you do this:

```cs
CopyCommand = ReactiveCommand.Create(async () =>
{
    var window = ((IClassicDesktopStyleApplicationLifetime)Application.Current.ApplicationLifetime).Windows[0];
    var clipboard = window.Clipboard;
    var dataObject = new DataObject();
    dataObject.Set(DataFormats.Text, AfterSql);
    await clipboard.SetDataObjectAsync(dataObject);
});
PasteCommand = ReactiveCommand.Create(async () =>
{
    var window = ((IClassicDesktopStyleApplicationLifetime)Application.Current.ApplicationLifetime).Windows[0];
    var clipboard = window.Clipboard;
    BeforeSql = await clipboard.GetTextAsync();

});
```

However the designer has no window, so you cannot reference any window from the designer. You can for example use `if(Design.IsDesignMode)` to exit the command if hit from designer.

**Reply by BlankWhite** [reply]
I made these changes, but it still doesn't resolve the issue ,I haven't clicked on anything

```
  CopyCommand = ReactiveCommand.Create(async () =>
  {
      if (!Design.IsDesignMode)
      {
          var window = ((IClassicDesktopStyleApplicationLifetime)Application.Current.ApplicationLifetime).Windows[0];
          var clipboard = window.Clipboard;
          var dataObject = new DataObject();
          dataObject.Set(DataFormats.Text, AfterSql);
          await clipboard.SetDataObjectAsync(dataObject);
      }
    
  });
  PasteCommand = ReactiveCommand.Create(async () =>
  {
      if (!Design.IsDesignMode)
      {
          var window = ((IClassicDesktopStyleApplicationLifetime)Application.Current.ApplicationLifetime).Windows[0];
          var clipboard = window.Clipboard;
          BeforeSql = await clipboard.GetTextAsync();
      }
    

  });
```
The error still persists.
```
08:28:44.217 [Error] 20908 UpdateXamlResult error
System.Xaml.XamlException: Dynamic code generation is not supported on this platform.
```

**Reply by BlankWhite** [reply, marked as answer]
I found the issue; uncommenting this line resolved the issue. It might be a bug.
`<PublishAot>true</PublishAot>`

**Reply by vpenades** [reply]
This my remove the error, but it's not a solution if you actually want to publish to AOT

**Reply by timunie** [reply]
The previewer can't work with the AOT output. So it can only work if you build it in DebugMode. 

Probably the aot switch can be made conditional if that alone doesn't help.

**Reply by vpenades** [reply]
As I understand it, the binaries of a publish are the final product, not intermediate files, so the designer should use the binaries located in the debug or release directories, not in the publish directory. Actually most of the time, the publish will use the binaries from the release directory.

Also for many, the whole point of using avalonia is the advertised "we support AOT" with all the compiled bindings and so on, which is great, so I think this is a loop hole that needs to be addressed by the designer, not by the end developers.

**Reply by timunie** [reply]
In the long term, yes. For now it's good there is a workaround available. In the background we are working on an improved previewer. Let's see if it also solves this bit.

**Comment by dabbinavo** [comment]
@BlankWhite @vpenades ,
You may use this as a workaround while keeping Aot publishing enabled:

```xml
  <PropertyGroup Condition="'$(_IsPublishing)' == 'true'">
    <PublishAot>true</PublishAot>
  </PropertyGroup>
  ```
